module/zfs/bpobj.c

```diff
@@ -206,29 +206,64 @@ bpobj_is_empty(bpobj_t *bpo)
 	    (!bpo->bpo_havesubobj || bpo->bpo_phys->bpo_num_subobjs == 0));
 }
 
+/*
+ * A recursive iteration of the bpobjs would be nice here but we run the risk
+ * of overflowing function stack space.  Instead, find each subobj and add it
+ * to the head of our list so it can be scanned for subjobjs.  Like a
+ * recursive implementation, the "deepest" subobjs will be freed first.
+ * When a subobj is found to have no additional subojs, free it.
+ */
+typedef struct bpobj_info {
+	bpobj_t *bpi_bpo;
+	struct bpobj_info *bpi_parent;
+	uint64_t bpi_object;
+	uint64_t bpi_index;
+	list_node_t bpi_node;
+} bpobj_info_t;
+
+static bpobj_info_t *
+bpi_alloc(bpobj_t *bpo, bpobj_info_t *parent, uint64_t obj, uint64_t index)
+{
+	bpobj_info_t *bpi = kmem_zalloc(sizeof (bpobj_info_t), KM_SLEEP);
+	bpi->bpi_bpo = bpo;
+	bpi->bpi_parent = parent;
+	bpi->bpi_object = obj;
+	bpi->bpi_index = index;
+	return (bpi);
+}
+
+/*
+ * Update bpobj and all of its parents with new space accounting.
+ */
+static void
+propagate_space_reduction(bpobj_info_t *bpi, uint64_t freed,
+    uint64_t comp_freed, uint64_t uncomp_freed, dmu_tx_t *tx)
+{
+
+	for (; bpi != NULL; bpi = bpi->bpi_parent) {
+		bpobj_t *p = bpi->bpi_bpo;
+		ASSERT(dmu_buf_is_dirty(p->bpo_dbuf, tx));
+		p->bpo_phys->bpo_bytes -= freed;
+		ASSERT3S(p->bpo_phys->bpo_bytes, >=, 0);
+		p->bpo_phys->bpo_comp -= comp_freed;
+		p->bpo_phys->bpo_uncomp -= uncomp_freed;
+	}
+
+}
+
 static int
-bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
-    boolean_t free)
+bpobj_iterate_blkptrs(bpobj_info_t *bpi, bpobj_itor_t func, void *arg,
+    dmu_tx_t *tx, boolean_t free)
 {
-	dmu_object_info_t doi;
-	int epb;
-	int64_t i;
 	int err = 0;
+	int64_t i;
+	uint64_t freed = 0, comp_freed = 0, uncomp_freed = 0;
 	dmu_buf_t *dbuf = NULL;
-
-	ASSERT(bpobj_is_open(bpo));
-	mutex_enter(&bpo->bpo_lock);
-
-	if (free)
-		dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+	bpobj_t *bpo = bpi->bpi_bpo;
 
 	for (i = bpo->bpo_phys->bpo_num_blkptrs - 1; i >= 0; i--) {
```

Comment:
since we are decrementing bpo_num_blkptrs as we go, I wonder if this should just be `while (bpo_num_blkptrs > 0)` (and then use `bpo_num_blkptrs - 1` rather than `i` to compute offset,blkoff)

Reply:
Looked at doing this but bpo_num_blkptrs is only decremented "if (free)".

Reply:
ah, good catch.